[WX-963] Unzip Engine Function

THWiseman opened the pull request:
Implemented the [unzip()](https://github.com/openwdl/wdl/blob/main/versions/1.1/SPEC.md#-pairarrayx-arrayy-unziparraypairx-y) engine function, new with WDL 1.1. 

As a new WDL 1.1 feature, it was implemented for `Biscayne` and `Cascades`, but not earlier versions (e.g. `draft-2`, `draft-3`).

aednichols reviewed (approved):
Nice!

jgainerdewar commented on wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/values/CascadesValueEvaluatorSpec.scala:
```diff
@@ -217,4 +217,30 @@ class CascadesValueEvaluatorSpec extends AnyFlatSpec with CromwellTimeoutSpec wi
+  }
+
+  it should "evaluate an unzip on an empty collection correctly" in {
+    val str = """ unzip([])"""
```
Is it worth testing `unzip([()])`?

jgainerdewar commented on wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/types/CascadesTypeEvaluatorSpec.scala:
```diff
@@ -64,4 +64,12 @@ class CascadesTypeEvaluatorSpec extends AnyFlatSpec with CromwellTimeoutSpec wit
       e.evaluateType(Map.empty) shouldBeValid WomArrayType(WomStringType)
     }
   }
+
```
There's another test in the Biscayne version, should it be added here?

jgainerdewar replied:
I think we're still missing a test here that is is Biscayne, no? Once that's in I think this is good to go!

jgainerdewar reviewed (commented):
Looks great! 

Reviewing this made me think about how we don't seem to have much unit testing of failure modes (my suffix PR suffered from the same issue) - what if I run a WDL with `unzip([("banana", 5), (6, "apple"])`? Is that valid, or does it fail because we can't find good types for the output arrays?

THWiseman commented:
> Looks great!
> 
> Reviewing this made me think about how we don't seem to have much unit testing of failure modes (my suffix PR suffered from the same issue) - what if I run a WDL with `unzip([("banana", 5), (6, "apple")])`? Is that valid, or does it fail because we can't find good types for the output arrays?

Agree with this sentiment! I added a few unit tests for basic failure modes.

To your more specific question: `unzip([("banana", 5), (6, "apple")])` will *succeed*, because putting stuff into arrays homogenizes the type, and the integers are coercible to strings. For that to fail, I think we would want arrays to throw errors instead of attempting to homogenize types, which is a big and breaking change that we probably shouldn't make. 

However, unzip `([(1, 2), ([1,2,3], 3)])` will fail because there's no way to coerce between `int` and `array[int]`. Added a test case to assert this.
